src/Documentation/Event-Sourcing/LogConsistencyProviders.md

```diff
@@ -0,0 +1,54 @@
+---
+layout: page
+title: Included Log-Consistency Providers
+---
+
+# Built-In Log-Consistency Providers
+
+The `Microsoft.Orleans.EventSourcing` package includes several log-consistency providers that cover basic scenarios suitable to get started, and allow some extensibility.
+
+
+### Orleans.EventSourcing.**StateStorage**.LogConsistencyProvider
+
+This provider stores *grain state snapshots*, using a standard storage provider that can be configured independently. 
+
+The data that is kept in storage is an object that contains both the grain state (as specified by the first type parameter to `JournaledGrain`) and some meta-data (the version number, amd a special tag that is used to avoid duplication of events when storage accesses fail).
+
+Since the entire grain state is read/written every time we access storage, this provider is not suitable for objects whose grain state is very large. 
+
+This provider does not support `RetrieveConfirmedEvents`: it cannot retrieve the events from storage because the events are not persisted.
```

Comment:
The initial draft of the event sourcing API made some restrictive assumptions 

1. Grains do only one thing with the events: use them to update the grain state. 
2. The grain state and transition function change very rarely, e.g. due to bug fixes; in all those situations, the grain is deactivated before the change

However, I think this is a bit too narrowly conceived. For example, in the following situations, a grain may need to access the event sequence (parts of it or all of it):

- the grain is maintaining some set of views that can change dynamically. 
- the grain wants to react to each event in some special way (in addition to just updating the grain state). For example, it may want to forward the event to some client.

Neither of these can be done nicely using just the initial draft of the API, but can be done very cleanly if we additionally support the `RetrieveConfirmedEvents`. For example, the grain code can handle `OnStateChanged`, and then use `RetrieveConfirmedEvents` to get access to new events in the sequence. It all works very robustly together (e.g. no surprises due to races, because we validate that grain code accesses only the confirmed part of the log, which does not change). 


It is not difficult to implement `RetrieveConfirmedEvents` in the provider: the requested segment of the event sequence either exists (in which case we can fetch and return it) or it doesn't (in which case we throw `NotSupportedException`, as documented). For all true event sourcing solutions, the log exists. For the `StateStorage` style of provider, it does not.

I don't think it is a terribly leaky abstraction, its abstraction level matches the rest of the API. 

Of course it is leaky in the sense that it can expose limitations of the provider (e.g. the provider may have thrown away the log, or may take a long time fetching it from cold storage). But I don't think this is something we should try to avoid at all cost... it seems not reasonable to me to expect that all providers work perfectly with all applications. If we require that, then all providers have to be super-heavily engineered. For example, it would be impossible to use light-weight providers that take snapshots immediately, all the time, and discard the events quickly.

About the snapshotting: we probably need a separate discussion on that, and write some higher-quality provider that actually does it. I never quite understood the details of how this is supposed to work. How does the provider know when it has to replay the log because the transitions have changed? After it fails to successfully deserialize the snapshot?